research-prototype/ipa/asort.py

```diff
@@ -0,0 +1,166 @@
+        break_point()
+        v = D.get_vector()
+        D.assign_slice_vector(idx, v)
+    break_point()
+    delshuffle(shuffle)
+
+
+def double_dest(bs):
+    """
+    bs is an n by 2 bit array.
+    """
+    num, _ = bs.sizes
+    # BUG this is never used
+    bits = sint.Array(num * 4)
+    col0 = bs.get_column(0)
+    col1 = bs.get_column(1)
+    prod = col0 * col1
+    # BUG this is never assigned
+    cum.assign_vector(prod - col0 - col1 + 1)  # 00
+    cum.assign_vector(col1 - prod, base=num)  # 01
+    cum.assign_vector(col0 - prod, base=2 * num)  # 10
+    cum.assign_vector(prod, base=3 * num)  # 11
+
+    @for_range(len(B) - 1)
+    def _(i):
+        cum[i + 1] = cum[i + 1] + cum[i]
+
+    one_contrib = cum.get_vector(size=num)
+    col0_contrib = (cum.get_vector(base=2 * num, size=num)
+                    - one_contrib)
+    col1_contrib = (cum.get_vector(base=num, size=num)
+                    - one_contrib)
+    prod_contrib = (one_contrib
+                    + cum.get_vector(base=3*num, size=num))
+    return (one_contrib
+            + col0 * col0_contrib
+            + col1 * col1_contrib
+            + prod * prod_contrib-1)
+
+
+def double_bit_radix_sort(bs, D):
+    """
+    Use two bits at a time.
+    There's an annoying problem if n_bits is odd.
+    """
+    n_bits, num = bs.sizes
+    h = Array.create_from(sint(regint.inc(num)))
+
+    # Test if n_bits is odd
+    @for_range(n_bits // 2)
+    def _(i):
+        perm = double_dest(bs[2 * i: 2 * i + 2])
+        reveal_sort(perm, h, reverse=False)
+
+        @if_e(2 * i + 3 < n_bits)
+        def _():  # sort the next 2 bits
+            reveal_sort(h, bs[2 * i + 2: 2 * i + 4], reverse=True)
+
+        @else_
+        def _():
+            @if_(n_bits % 2 == 1)
+            def odd_case():
+                reveal_sort(h, bs[-1], reverse=True)
+                c = Array.create_from(dest_comp(bs[-1]))
+                reveal_sort(c, h, reverse=False)
+    # Now take care of the odd case
+    reveal_sort(h, D, reverse=True)
+
+
+def bit_radix_sort(bs, D):
+    n_bits, num = bs.sizes
+    B = sint.Matrix(num, 2)
+    h = Array.create_from(sint(regint.inc(num)))
+
+    @for_range(n_bits)
+    def _(i):
+        b = bs[i]
+        B.set_column(0, 1 - b.get_vector())
+        B.set_column(1, b.get_vector())
+        c = Array.create_from(dest_comp(B))
+        reveal_sort(c, h, reverse=False)
+
+        @if_e(i < n_bits - 1)
+        def _():
+            reveal_sort(h, bs[i + 1], reverse=True)
+
+        @else_
+        def _():
+            reveal_sort(h, D, reverse=True)
+
+
+def radix_sort(k, D, n_bits=None, signed=True):
+    assert len(k) == len(D)
+    bs = Matrix.create_from(k.get_vector().bit_decompose(n_bits))
+    if signed and len(bs) > 1:
+        bs[-1][:] = bs[-1][:].bit_not()
+    bit_radix_sort(bs, D)
+
+
+def two_bit_radix_sort(k, D, n_bits):
+    # BUG: radix sort doesn't accept two_bit...
+    return radix_sort(k, D, n_bits, two_bit=True)
```

Comment:
`two_bit` isn't a vailid kwarg for `radix_sort`. should it be `signed`? if not, we need to fix.